vm based providers: Introduce robust stack provider

oshoval opened the pull request:
The following provider supports both dual stack and single stack IPv6.
Selected during `cluster-up`, according `KUBEVIRT_SINGLE_STACK` value (default: false - dual stack).

For now done only for 1.25, once it is ready, will adapt 1.26 as well.

Additional changes:
1. Delete k8s-1.25-ipv6 provider (once this PR is bumped, the new provider must be used).
See https://github.com/kubevirt/project-infra/pull/2580
2. Add missing psa configuration for kubeadm_ipv6.conf
3. Update publish script to stop creating the single stack provider.
4. Update cluster-up to sanity check in case of single stack (check that the pods have only IPv6)

oshoval commented:
check-provision-k8s-1.25-ipv6 expected to fail because it is deleted, will remove the lane in a dedicated PR once
this PR is reviewed

oshoval commented:
/test check-provision-k8s-1.25

oshoval commented:
/test check-provision-k8s-1.25

oshoval commented:
/test check-provision-k8s-1.25

oshoval commented:
k8s-12.26 not related
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirtci/954/check-provision-k8s-1.26/1620025574892244992

```
+ dnf install -y cri-o
Stable Releases of Upstream github.com/containe 1.0 MB/s |  48 kB     00:00    
devel:kubic:libcontainers:stable:cri-o:1.26 (Ce 4.5 kB/s | 636  B     00:00    
Errors during downloading metadata for repository 'devel_kubic_libcontainers_stable_cri-o_1.26':
  - Status code: 404 for https://storage.googleapis.com/kubevirtci-crio-mirror/devel_kubic_libcontainers_stable_cri-o_1.26/repodata/ae28764aea2f52436da133f06150efd68565a32bf21b47ef1bc37412567bfcfd-primary.xml.gz (IP: 142.250.114.128)
  - Status code: 404 for https://storage.googleapis.com/kubevirtci-crio-mirror/devel_kubic_libcontainers_stable_cri-o_1.26/repodata/525ee0a085db4622cd57caddcd5ab6e2ef2b0ebaec42997df7bc8bcb052f6bda-filelists.xml.gz (IP: 142.250.114.128)
Error: Failed to download metadata for repo 'devel_kubic_libcontainers_stable_cri-o_1.26': Yum repo downloading error: Downloading error(s): repodata/ae28764aea2f52436da133f06150efd68565a32bf21b47ef1bc37412567bfcfd-primary.xml.gz - Cannot download, all mirrors were already tried without success; repodata/525ee0a085db4622cd57caddcd5ab6e2ef2b0ebaec42997df7bc8bcb052f6bda-filelists.xml.gz - Cannot download, all mirrors were already tried without success
```

oshoval commented:
/test check-provision-k8s-1.25

a flake that isn't related it seem (saw it on the past as well)
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirtci/954/check-provision-k8s-1.25/1620025574799970304

brianmcarey commented on cluster-provision/k8s/1.25/manifests/kubeadm_ipv6.conf:
```diff
@@ -26,7 +26,11 @@ apiServer:
     enable-admission-plugins: NamespaceLifecycle,LimitRanger,ServiceAccount,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,MutatingAdmissionWebhook,ValidatingAdmissionWebhook,ResourceQuota
     feature-gates: AdvancedAuditing=true
     bind-address: "::"
+    admission-control-config-file: /etc/kubernetes/psa.yaml
```
I am not sure if we are enabling PSA on providers earlier than 1.26? @acardace - am I right here?

oshoval replied:
we do on 1.25 and it was missing
https://github.com/kubevirt/kubevirtci/pull/915/files#r1052081728

brianmcarey replied:
Ok thanks @oshoval

oshoval replied:
you mean enabled using `KUBEVIRT_PSA` right ?
`kubeadm.conf` changes do exists on 1.25 as well

brianmcarey replied:
@oshoval  I guess if a user does the following 
`export KUBEVIRT_SINGLE_STACK=true KUBEVIRT_PROVIDER=k8s-1.24`

They get PSA configuration for a 1.24 cluster?

oshoval replied:
indeed not support for 1.24
moreover the feature determines only if its a single stack IPv6 or dual stack (default)

brianmcarey reviewed (commented):
Thanks Or, it will be great to reduce the number of providers that we are publishing. Just one question on PSA.

brianmcarey reviewed (commented):
/lgtm

oshoval commented:
Added the last commit which remove the stack type provision parameter
we don't need it anymore

brianmcarey commented:
/override check-provision-k8s-1.26

Changes applied to 1.25 provider.

brianmcarey reviewed (commented):
/lgtm

oshoval commented:
Changes:
Added latest commit to support 1.26 as well (centos8 one)
Removed the workaround that was on `cluster-provision/k8s/check-cluster-up.sh` (because now we do support robust stack also on 1.26)
Updated `cluster-up/up.sh` with cosmetics

oshoval commented:
Fix check-cluster-up `find` command (had to be relative to the current directory which is `cluster-provision/k8s`)

oshoval commented:
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirtci/954/check-provision-k8s-1.26/1625758534794743808

due to known PSA problem right ?

oshoval commented:
Added support for `k8s-1.26-centos9`

brianmcarey commented:
/test check-provision-k8s-1.26-centos9

oshoval commented:
> /test check-provision-k8s-1.26-centos9

`dnsmasq: failed to create inotify: Too many open files`

we had it in the past, leakage of descriptors on the machine?
or need to config the limit (this what helped in the past IIRC)

brianmcarey commented:
/retest-required

brianmcarey commented:
/test check-provision-k8s-1.26-centos9

brianmcarey reviewed (approved):
/approve

brianmcarey commented:
/override check-provision-k8s-1.26
